README.md
# Chrono

[![Build Status](http://img.shields.io/travis/icecave/chrono/master.svg?style=flat-square)](https://travis-ci.org/icecave/chrono)
[![Code Coverage](https://img.shields.io/codecov/c/github/icecave/chrono/master.svg?style=flat-square)](https://codecov.io/github/icecave/chrono)
[![Latest Version](http://img.shields.io/packagist/v/icecave/chrono.svg?style=flat-square&label=semver)](https://semver.org)

**Chrono** is a PHP date & time library that is decoupled from the system clock.

    composer require icecave/chrono

## Rationale

Many date & time operations in the core PHP libraries require access to system state such as the current wall time, or
resources such as timezone databases. These hard-wired dependencies can make it very difficult to write well-abstracted
and testable code when dealing with time-sensitive operations.

**Chrono** provides a set of date & time classes that are completely decoupled from the system and hence behave
consistently, regardless of system state and configuration (such as the
[date.timezone INI directive](http://www.php.net/manual/en/datetime.configuration.php#ini.date.timezone)).

A [SystemClock](src/Clock/SystemClock.php) instance must be explicitly constructed before any global
date & time operations are used. Classes that require use of a clock may take a [ClockInterface](src/Clock/ClockInterface.php)
as a [dependency](http://en.wikipedia.org/wiki/Dependency_injection), improving decoupling and testability.

## Concepts

* [Clock](src/Clock/ClockInterface.php): A factory for chronological measurements.
* [Time](src/TimeInterface.php): A chronological measurement with a time component.
* [Date](src/DateInterface.php): A chronological measurement with a date component.
* [Time Point](src/TimePointInterface.php): A discreet point on the time-continuum.
* [Time Span](src/TimeSpan/TimeSpanInterface.php): An un-anchored span of time.
* [Interval](src/Interval/IntervalInterface.php): A span of time between two *Time Points*.

## Implementations

* [System Clock](src/Clock/SystemClock.php): A factory for chronological measurements that uses the system clock.
* [Test Clock](src/Clock/TestClock.php): A clock that can be manually manipulated for testing purposes.
* [Date](src/Date.php): Represents a date. Models the *Time Point* and *Date* concepts.
* [Time of Day](src/TimeOfDay.php): Represents a time of day. Models the *Time* concept.
* [Date Time](src/DateTime.php): Represents a time of day on specific date. Models the *Time Point*, *Date* and *Time* concepts.
* [Interval](src/Interval/Interval.php): A span of time between two *Time Points*. Models the *Interval* concept.
* [Month](src/Interval/Month.php): A one month time span. Models the *Interval* concept.
* [Year](src/Interval/Year.php): A one year time span. Models the *Interval* concept.
* [Duration](src/TimeSpan/Duration.php): A time span measured in seconds with no beginning or end. Models the *Time Span* concept.
* [Period](src/TimeSpan/Period.php): A time span specified in component form (eg: 3 months, 4 days), models the *Time Span* concept.

## Examples

### Getting the current time

In order to get the current time you need to use a [clock](src/Clock/ClockInterface.php). Most of the time in production
code you will use the [SystemClock](src/Clock/SystemClock.php) class, which uses the machine's current system time and
time zone information.

```php
use Icecave\Chrono\Clock\SystemClock;

// Construct a new system clock ...
$clock = new SystemClock;

// Obtain a DateTime instance representing the current date and time ...
$now = $clock->localDateTime();

// Obtain a Date instance representing the current date ...
$today = $clock->localDate();

// Obtain the current time of day ...
$timeOfDay = $clock->localTime();
```

Each of the clock methods shown above has a [UTC](http://en.wikipedia.org/wiki/Coordinated_Universal_Time) counterpart.
For example, to obtain the current time in UTC you can use the following code:

```php
$nowUtc = $clock->utcDateTime();
```

### String formatting

To produce a formatted string representing a [Date](src/Date.php), [DateTime](src/DateTime.php),
[TimeOfDay](src/TimeOfDay.php) or [TimeZone](src/TimeZone.php) instance use the `format()`
method.

The output is specified using the same format as PHP's [built-in date() function](http://php.net/manual/en/function.date.php).

```php
$now = $clock->localDateTime();
$string = $now->format('Y-m-d H:i:s');
```

Casting the object as a string (or calling `isoString()`) produces an [ISO-8601](http://en.wikipedia.org/wiki/ISO_8601)
string representation.

### Unix timestamps

[Date](src/Date.php) and [DateTime](src/DateTime.php) instances can be produced from unix
timestamps using the `fromUnixTime()` static method. The unix timestamp can be retrieved using `unixTime()`.

```php
$dateTime = DateTime::fromUnixTime(1367823963);
$timestamp = $dateTime->unixTime();
```

### PHP native "DateTime" objects

[Date](src/Date.php) and [DateTime](src/DateTime.php) instances can be produced from
native PHP [DateTime](http://php.net/manual/en/class.datetime.php) instances using the `fromNativeDateTime()` static
method, and can be converted to a native DateTime using `nativeDateTime()`.

```php
use DateTime as NativeDateTime;
use Icecave\Chrono\DateTime;

$dateTime = DateTime::fromNativeDateTime(new NativeDateTime);
$nativeDateTime = $dateTime->nativeDateTime();
```


CHANGELOG.md
# Chrono Changelog

### 2.0.0 (2020-08-25)

- **[BC]** Bump minimum PHP version to 7.2
- **[BC]** Remove `PackageInfo` class
- **[FIX]** `Date::toTimeZone()` now always returns a `DateTime`, even if "converting" to the same timezone

### 1.0.4 (2016-02-19)

- **[IMPROVED]** Added support for partial second parsing in internal class `Iso8604`, in preparation for adding partial-second support to the public interface

### 1.0.3 (2015-02-16)

- **[IMPROVED]** `DateTime` and `TimeOfDay` ISO string parsing now allows (but discards) partial-second information

### 1.0.2 (2014-11-10)

- **[FIXED]** `DateTime::fromNativeDateTime()` correctly preserves DST flag when converting from timezones that use abbreviated names (e.g. PST, PDT, etc)

### 1.0.1 (2014-10-20)

- **[IMPROVED]** Allow either version 2.* or 3.* of [icecave/isolator](https://github.com/isolator)

### 1.0.0 (2014-09-09)

- **[NEW]** Added convenience factory methods on `Duration`

### 0.4.0 (2014-02-27)

- **[NEW]** Added `TestClock`, a clock implementation that can be manipulated manually to ease testing of time-dependent code
- **[IMPROVED]** Updated autoloader to [PSR-4](http://www.php-fig.org/psr/psr-4/)

### 0.3.0 (2014-01-24)

- **[BC]** Methods names on `TimeInterface` are now singular (ie, `hour()` vs `hours()`)
- **[BC]** `Interval::isoStringWithDuration()` is now more correctly named `isoStringWithPeriod()`
- **[FIXED]** `SystemClock::unixTime()` now properly suspends when the clock is suspended
- **[IMPROVED]** `Period::resolveToInterval()` and `Duration::resolveToInterval()` can now handle negative time spans
- **[NEW]** Added ISO-8601 parsing to `Period`, `Interval`, `Year` and `Month`
- **[NEW]** Added `DateInterface::isoDayOfWeek()` and `numericDayOfWeek()`
- **[NEW]** Added methods for converting `Period` and `Duration` to/from PHP's `DateInterval` type
- **[NEW]** Added `ClockInterface::unixTimeAsFloat()`
- **[NEW]** Added `ClockInterface::sleep()` and `sleepUntil()`
- **[NEW]** Added `TimerInterface`, `Timer` and `ClockInterface::createTimer()`
- **[NEW]** Implemented several iterator types for iterating over time spans (see [#3](https://github.com/icecave/chrono/issues/3))
- **[NEW]** Most objects can now be compared using [Parity](https://github.com/icecave/parity)

### 0.2.0 (2013-05-13)

- **[BC]** Moved `Duration` and `TimeSpanInterface` into `TimeSpan` namespace
- **[BC]** Removed `AbstractDuration` and `DurationInterface`
- **[BC]** Renamed `TimeSpanInterface::resolve()` to `resolveToSeconds()`
- **[IMPROVED]** `TimePointInterface::add()` and `subtract()` now accept raw seconds as integer, as well as a time span
- **[FIXED]** Added missing time zone support to `Date::fromNativeDateTime()` and `nativeDateTime()`
- **[NEW]** Implemented `Period`, a model of `TimeSpanInterface`
- **[NEW]** Added `TimeSpanInterface::inverse()`, `isEmpty()`, `__toString()` and `resolveTo***()` methods
- **[NEW]** Added `TimePointInterface::differenceAsSeconds()` and `differenceAsPeriod()`
- **[NEW]** Added `is[Not]EqualTo()` and `is[Less/Greater]Than[OrEqualTo]()` to all types with `compare()`
- **[NEW]** Added `Year::fromTimePoint()` and `Month::fromTimePoint()`

### 0.1.1 (2013-05-09)

- **[FIXED]** Fixed inverted comparison of time zone in `DateTime/Date/TimeOfDay::compare()`

### 0.1.0 (2013-05-06)

- Initial release


src/Detail/Iso8601.php
<?php

namespace Icecave\Chrono\Detail;

use InvalidArgumentException;

/**
 * ISO-8601 standard parsing and formatting functions.
 *
 * See link for the full specifics on formats.
 * @link http://en.wikipedia.org/wiki/ISO_8601
 */
abstract class Iso8601
{
    /**
     * @param int $year
     *
     * @return string A string representing the data in an ISO compatible year format (YYYY).
     */
    public static function formatYear($year)
    {
        return sprintf('%04d', $year);
    }

    /**
     * @param int $year
     * @param int $month
     *
     * @return string A string representing the data in an ISO compatible year month format (YYYY-MM).
     */
    public static function formatYearMonth($year, $month)
    {
        return sprintf('%04d-%02d', $year, $month);
    }

    /**
     * @param int $year
     * @param int $month
     * @param int $day
     *
     * @return string A string representing the data in an ISO compatible date format (YYYY-MM-DD).
     */
    public static function formatDate($year, $month, $day)
    {
        return sprintf(
            '%04d-%02d-%02d',
            $year,
            $month,
            $day
        );
    }

    /**
     * @param int    $hour
     * @param int    $minute
     * @param int    $second
     * @param string $timeZone Time zone in ISO compatible time zone format.
     *
     * @return string A string representing the data in an ISO compatible time format (hh:mm:ss[+-]hh:mm).
     */
    public static function formatTime($hour, $minute, $second, $timeZone)
    {
        return sprintf(
            '%02d:%02d:%02d%s',
            $hour,
            $minute,
            $second,
            $timeZone
        );
    }

    /**
     * @param int    $year
     * @param int    $month
     * @param int    $day
     * @param int    $hour
     * @param int    $minute
     * @param int    $second
     * @param string $timeZone Time zone in ISO compatible time zone format.
     *
     * @return string A string representing the data in an ISO compatible date time format (YYYY-MM-DDThh:mm:ss[+-]hh:mm).
     */
    public static function formatDateTime($year, $month, $day, $hour, $minute, $second, $timeZone)
    {
        return sprintf(
            '%04d-%02d-%02dT%02d:%02d:%02d%s',
            $year,
            $month,
            $day,
            $hour,
            $minute,
            $second,
            $timeZone
        );
    }

    /**
     * @param int $offset The offset from UTC in seconds.
     *
     * @return string A string representing the data in an ISO compatible time zone format ([+-]hh:mm).
     */
    public static function formatTimeZone($offset)
    {
        $seconds = abs($offset);
        $minutes = ($seconds % 3600) / 60;
        $hours   = $seconds / 3600;

        return sprintf(
            '%s%02d:%02d',
            $offset >= 0 ? '+' : '-',
            $hours,
            $minutes
        );
    }

    /**
     * @param int $years
     * @param int $months
     * @param int $days
     * @param int $hours
     * @param int $minutes
     * @param int $seconds
     *
     * @return string A string representing the data in an ISO compatible duration format (PnYnMnDTnHnMnS).
     */
    public static function formatDuration($years, $months, $days, $hours, $minutes, $seconds)
    {
        $dateParts = '';
        if ($years !== 0) {
            $dateParts .= $years . 'Y';
        }
        if ($months !== 0) {
            $dateParts .= $months . 'M';
        }
        if ($days !== 0) {
            $dateParts .= $days . 'D';
        }

        $timeParts = '';
        if ($hours !== 0) {
            $timeParts .= $hours . 'H';
        }
        if ($minutes !== 0) {
            $timeParts .= $minutes . 'M';
        }
        if ($seconds !== 0 || (strlen($dateParts) === 0 && strlen($timeParts) === 0)) {
            $timeParts .= $seconds . 'S';
        }

        if (strlen($timeParts) > 0) {
            $timeParts = 'T' . $timeParts;
        }

        return 'P' . $dateParts . $timeParts;
    }

    /**
     * Standard year format:
     *   YYYY
     *
     * @link http://en.wikipedia.org/wiki/ISO_8601#Calendar_dates
     *
     * Note: Negative years (BC) are not supported.
     *
     * @param string $isoString A string containing a year in any ISO-8601 compatible year format.
     *
     * @return int The year number from the ISO compatible string.
     */
    public static function parseYear($isoString)
    {
        $matches = [];
        if (preg_match(self::YEAR, $isoString, $matches) === 1) {
            $year = intval($matches[1]);
        } else {
            throw new InvalidArgumentException('Invalid ISO year: "' . $isoString . '".');
        }

        return $year;
    }

    /**
     * Standard year month format:
     *   YYYY-MM
     *
     * @link http://en.wikipedia.org/wiki/ISO_8601#Calendar_dates
     *
     * @param string $isoString A string containing a year month in any ISO-8601 compatible year month format.
     *
     * @return array<string,integer> The year month periods from the ISO compatible string.
     */
    public static function parseYearMonth($isoString)
    {
        $parts = [
            'year'   => 0,
            'month'  => 0,
        ];

        $matches = [];
        if (preg_match(self::YEAR_MONTH, $isoString, $matches) === 1) {
            $parts['year']  = intval($matches[1]);
            $parts['month'] = intval($matches[2]);
        } else {
            throw new InvalidArgumentException('Invalid ISO year month: "' . $isoString . '".');
        }

        return $parts;
    }

    /**
     * Standard date formats:
     *   YYYY-MM-DD[timezone]
     *   YYYYMMDD[timezone]
     *
     * @link http://en.wikipedia.org/wiki/ISO_8601#Calendar_dates
     *
     * Note: Formats YYYY-MM and YYYYMM for reduced precision are currently not supported.
     *
     * @param string $isoString A string containing a date in any ISO-8601 compatible date format.
     *
     * @return array<string,integer|null> The date periods from the ISO compatible string.
     */
    public static function parseDate($isoString)
    {
        $date = [
            'year'   => 0,
            'month'  => 0,
            'day'    => 0,
            'offset' => null,
        ];

        $matches = [];
        if (preg_match(self::DATE_EXTENDED, $isoString, $matches) === 1 ||
            preg_match(self::DATE_BASIC, $isoString, $matches) === 1) {
            $date['year']  = intval($matches[1]);
            $date['month'] = intval($matches[2]);
            $date['day']   = intval($matches[3]);

            if (count($matches) > 4 && strlen($matches[4]) > 0) {
                $date['offset'] = self::parseTimeZone($matches[4]);
            }
        } else {
            throw new InvalidArgumentException('Invalid ISO date: "' . $isoString . '".');
        }

        return $date;
    }

    /**
     * Standard time formats:
     *   hh:mm:ss[timezone]
     *   hhmmss[timezone]
     *
     * @link http://en.wikipedia.org/wiki/ISO_8601#Times
     *
     * Note: Formats hh:mm and hhmm for reduced precision are currently not supported.
     *
     * @param string $isoString A string containing a time in any ISO-8601 compatible time format.
     *
     * @return array<string,integer|null> The time periods from the ISO compatible string.
     */
    public static function parseTime($isoString)
    {
        $time = [
            'hour'        => 0,
            'minute'      => 0,
            'second'      => 0,
            'microsecond' => 0,
            'offset'      => null,
        ];

        $matches = [];
        if (preg_match(self::TIME_EXTENDED, $isoString, $matches) === 1 ||
            preg_match(self::TIME_BASIC, $isoString, $matches) === 1) {
            $time['hour']   = intval($matches[1]);
            $time['minute'] = intval($matches[2]);
            $time['second'] = intval($matches[3]);
            $time['microsecond'] = intval(str_pad(substr($matches[4], 0, 6), 6, '0'));

            if (count($matches) > 5 && strlen($matches[5]) > 0) {
                $time['offset'] = self::parseTimeZone($matches[5]);
            }
        } else {
            throw new InvalidArgumentException('Invalid ISO time: "' . $isoString . '".');
        }

        return $time;
    }

    /**
     * Standard date time formats:
     *   <date>T<time>[timezone]
     *   <date> <time>[timezone]
     *
     * @link http://en.wikipedia.org/wiki/ISO_8601#Combined_date_and_time_representations
     *
     * @param string $isoString A string containing a date time in any ISO-8601 compatible date time format.
     *
     * @return array<string,integer|null> The date time periods from the ISO compatible string.
     */
    public static function parseDateTime($isoString)
    {
        $dateTime = [
            'year'        => 0,
            'month'       => 0,
            'day'         => 0,
            'hour'        => 0,
            'minute'      => 0,
            'second'      => 0,
            'microsecond' => 0,
            'offset'      => null,
        ];

        $matches = [];
        if (preg_match(self::DATETIME_EXTENDED, $isoString, $matches) === 1 ||
            preg_match(self::DATETIME_BASIC, $isoString, $matches) === 1) {
            $dateTime['year']   = intval($matches[1]);
            $dateTime['month']  = intval($matches[2]);
            $dateTime['day']    = intval($matches[3]);

            $dateTime['hour']   = intval($matches[4]);
            $dateTime['minute'] = intval($matches[5]);
            $dateTime['second'] = intval($matches[6]);
            $dateTime['microsecond'] = intval(str_pad(substr($matches[7], 0, 6), 6, '0'));

            if (count($matches) > 8 && strlen($matches[8]) > 0) {
                $dateTime['offset'] = self::parseTimeZone($matches[8]);
            }
        } else {
            throw new InvalidArgumentException('Invalid ISO date time: "' . $isoString . '".');
        }

        return $dateTime;
    }

    /**
     * Standard time zone formats:
     *   Z
     *   +/-hh
     *   +/-hhmm
     *   +/-hh:mm
     *
     * @link http://en.wikipedia.org/wiki/ISO_8601#Time_zone_designators
     *
     * @param string $isoString A string containing a time zone in any ISO-8601 compatible time zone format, with the exception of allowing negative zero's.
     *
     * @return int The time zone offset from the ISO compatible string.
     */
    public static function parseTimeZone($isoString)
    {
        $offset = 0;

        $matches = [];
        if (preg_match(self::TIMEZONE_UTC, $isoString, $matches) === 1) {
            $offset = 0;
        } elseif (preg_match(self::TIMEZONE_OFFSET, $isoString, $matches) === 1) {
            $sign   = trim($matches[1]);
            $hour   = intval($matches[2]);
            $minute = count($matches) > 4 ? intval($matches[4]) : 0;

            $offset = intval(($hour * 60 * 60) + ($minute * 60));
            if ($sign === '-') {
                $offset = -$offset;
            }
        } else {
            throw new InvalidArgumentException('Invalid ISO time zone: "' . $isoString . '".');
        }

        return $offset;
    }

    /**
     * Standard duration formats:
     *   PnYnMnDTnHnMnS (zero periods may be ommitted)
     *   PnW
     *   P<date>T<time>
     *
     * @link http://en.wikipedia.org/wiki/ISO_8601#Durations
     *
     * Note: Decimal fractions are currently not supported.
     *
     * @param string $isoString A string containing a duration in any ISO-8601 compatible duration format.
     *
     * @return array<string,integer> The duration periods from the ISO compatible string.
     */
    public static function parseDuration($isoString)
    {
        if (!strlen($isoString) || $isoString === 'P' || substr($isoString, -1) === 'T') {
            throw new InvalidArgumentException('Invalid ISO duration: "' . $isoString . '".');
        }

        $duration = [
            'years'   => 0,
            'months'  => 0,
            'days'    => 0,
            'hours'   => 0,
            'minutes' => 0,
            'seconds' => 0,
        ];

        $matches = [];
        if (preg_match(self::DURATION_WEEK, $isoString, $matches)) {
            $duration['days'] = intval($matches[1]) * 7;
        } elseif (preg_match(self::DURATION_PERIOD, $isoString, $matches)) {
            $defaults = [
                'years'   => 0,
                'months'  => 0,
                'days'    => 0,
                'hours'   => 0,
                'minutes' => 0,
                'seconds' => 0,
            ];
            $matches += $defaults;

            $duration['years']   = intval($matches['years']);
            $duration['months']  = intval($matches['months']);
            $duration['days']    = intval($matches['days']);

            $duration['hours']   = intval($matches['hours']);
            $duration['minutes'] = intval($matches['minutes']);
            $duration['seconds'] = intval($matches['seconds']);
        } elseif (
            preg_match(self::DURATION_DATETIME_BASIC, $isoString, $matches) === 1 ||
            preg_match(self::DURATION_DATETIME_EXTENDED, $isoString, $matches) === 1
        ) {
            $duration['years']   = intval($matches[1]);
            $duration['months']  = intval($matches[2]);
            $duration['days']    = intval($matches[3]);

            $duration['hours']   = intval($matches[4]);
            $duration['minutes'] = intval($matches[5]);
            $duration['seconds'] = intval($matches[6]);

            // ISO-8601 Durations: Individual date and time values cannot exceed their moduli (e.g. a value of 13 for the month or 25 for the hour would not be permissible).
            if ($duration['months'] > 12 ||
                $duration['days'] > 31 ||
                $duration['hours'] > 24 ||
                $duration['minutes'] > 59 ||
                $duration['seconds'] > 59) {
                throw new InvalidArgumentException('Invalid ISO duration: "' . $isoString . '".');
            }
        } else {
            throw new InvalidArgumentException('Invalid ISO duration: "' . $isoString . '".');
        }

        return $duration;
    }

    /**
     * Standard interval formats:
     *   <start datetime>/<end datetime>
     *   <start datetime>/<duration>
     *   <duration>/<end datetime>
     *   <duration>
     *
     * @link http://en.wikipedia.org/wiki/ISO_8601#Time_intervals
     *
     * Note: Duration only format is not supported.
     *
     * @param string $isoString A string containing an interval in any ISO-8601 compatible interval format.
     *
     * @return array<string,array<string|array>> A associative array of interval type to interval parts from the ISO compatible string.
     */
    public static function parseInterval($isoString)
    {
        if (!strlen($isoString) || $isoString === '/') {
            throw new InvalidArgumentException('Invalid ISO interval: "' . $isoString . '".');
        }

        $parts = explode('/', $isoString, 2);

        if (count($parts) === 1) {
            // Duration only format is not supported.
            throw new InvalidArgumentException('Invalid ISO interval: "' . $isoString . '".');
        }

        list($firstPart, $secondPart) = $parts;

        // If $firstPart begins with P, than its a duration/end-datetime interval.
        if (strlen($firstPart) > 0 && $firstPart[0] === 'P') {
            return [
                'type'     => 'duration/datetime',
                'interval' => [$firstPart, $secondPart],
            ];
        }

        // If $secondPart begins with P, than its a datetime/duration interval.
        if (strlen($secondPart) > 0 && $secondPart[0] === 'P') {
            return [
                'type'     => 'datetime/duration',
                'interval' => [$firstPart, $secondPart],
            ];
        }

        // Only option left is a start-datetime/end-datetime interval.
        return [
            'type'     => 'datetime/datetime',
            'interval' => [$firstPart, $secondPart],
        ];
    }

    const YEAR       = '/^(\d\d\d\d)$/';
    const YEAR_MONTH = '/^(\d\d\d\d)-(\d\d)$/';

    const DATE_BASIC    = '/^(\d\d\d\d)(\d\d)(\d\d)(.*)$/';
    const DATE_EXTENDED = '/^(\d\d\d\d)-(\d\d)-(\d\d)(.*)$/';

    const TIME_BASIC    = '/^(\d\d)(\d\d)(\d\d)(?:\.(\d+))?(.*)$/';
    const TIME_EXTENDED = '/^(\d\d):(\d\d):(\d\d)(?:\.(\d+))?(.*)$/';

    const DATETIME_BASIC    = '/^(\d\d\d\d)(\d\d)(\d\d)[T| ](\d\d)(\d\d)(\d\d)(?:\.(\d+))?(.*)$/';
    const DATETIME_EXTENDED = '/^(\d\d\d\d)-(\d\d)-(\d\d)[T| ](\d\d):(\d\d):(\d\d)(?:\.(\d+))?(.*)$/';

    const TIMEZONE_UTC    = '/^(Z)$/';
    const TIMEZONE_OFFSET = '/^([+-])(\d\d)(:?(\d\d))?$/';

    const DURATION_WEEK              = '/^P(\d+)W$/';
    const DURATION_PERIOD            = '/^P(?:(?:(?P<years>\d+)Y)?(?:(?P<months>\d+)M)?(?:(?P<days>\d+)D)?(?:T(?:(?P<hours>\d+)H)?(?:(?P<minutes>\d+)M)?(?:(?P<seconds>\d+)S)?)?)?$/';
    const DURATION_DATETIME_BASIC    = '/^P(\d\d\d\d)(\d\d)(\d\d)T(\d\d)(\d\d)(\d\d)$/';
    const DURATION_DATETIME_EXTENDED = '/^P(\d\d\d\d)-(\d\d)-(\d\d)T(\d\d):(\d\d):(\d\d)$/';
}
